cmflib/contrib/graph_api.py

```diff
@@ -0,0 +1,583 @@
+        return [_graph_node(Stage, self._db, ctx, {"_pipeline": self}) for ctx in stage_contexts]
+
+    def executions(self) -> t.List["Execution"]:
+        """Return list of all executions in this pipeline"""
+        executions: t.List[Execution] = []
+        for stage in self.stages():
+            executions.extend(stage.executions())
+        return executions
+
+    def artifacts(self) -> t.List["Artifact"]:
+        """Return list of all unique artifacts consumed and produced by executions of this pipeline."""
+        artifacts: t.List[Artifact] = []
+        for execution in self.executions():
+            artifacts.extend(execution.inputs())
+            artifacts.extend(execution.outputs())
+        return unique(artifacts, "id")
+
+
+class Stage(Base):
+    """Class that represents pipeline stages by wrapping the `Context` concept in MLMD.
+
+    Users should not create instances of this class - use `MlmdGraph` class instead or other node wrappers.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._pipeline: t.Optional[Pipeline] = None
+        """Parent pipeline (lazily initialized)."""
+
+    @property
+    def pipeline(self) -> Pipeline:
+        """Return parent pipeline."""
+        if self._pipeline is None:
+            self._pipeline = _graph_node(
+                Pipeline, self._db, one(self._db.store.get_parent_contexts_by_context(context_id=self.id))
+            )
+        return self._pipeline
+
+    def executions(self) -> t.List["Execution"]:
+        """Return list of all executions of this stage."""
+        # noinspection PyProtectedMember
+        executions: t.List[metadata_store_pb2.Execution] = self._db._get_executions(stage_id=self.id)
+        return [_graph_node(Execution, self._db, execution, {"_stage": self}) for execution in executions]
+
+    def artifacts(self) -> t.List["Artifact"]:
+        """Return list of unique artifacts consumed and produced by executions of this stage."""
+        artifacts: t.List[Artifact] = []
+        for execution in self.executions():
+            artifacts.extend(execution.inputs())
+            artifacts.extend(execution.outputs())
+        return unique(artifacts, "id")
+
+
+class Execution(Base):
+    """Class that represents stage executions wrapping the `Execution` concept in MLMD.
+
+    Users should not create instances of this class - use `MlmdGraph` class instead or other node wrappers.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._stage: t.Optional[Stage] = None
+        """Stage for this execution (lazily initialized)."""
+
+    @property
+    def stage(self) -> Stage:
+        """Return stage of this execution."""
+        if self._stage is None:
+            self._stage = _graph_node(
+                Stage, self._db, one(self._db.store.get_contexts_by_execution(execution_id=self.id))
```

Comment:
@sergey-serebryakov Just a clarification, when calling "get_context_by_execution" will it return both the stage context and Pipeline context ?

Reply:
why do we need "one" function here ?